netinfo: add peer services column and outbound-only option

jonatack opened the pull request:
Been using this since May 2023.

- add a peer services column (considered displaying the p2p_v2 flag as "p" or "2"; proposing "2" here for continuity with the "v" column, but "p" is fine for me as well)
- clarify in the help that "relaytxes" and "addr_relay_enabled" are from getpeerinfo
- hoist (and rename) the max level constant to use in top-level help, to avoid overlooking to update the top-level help if the value of the constant changes (as caught by Larry Ruane in review below)
- add an optional "outonly" (or "o") argument for an outbound-only peer list, as suggested by Vasil Dimov in his review below. Several people have requested this, to keep the output within screen limits when running netinfo as a live dashboard (i.e. with `watch`) on a node with many peers. While doing this, also permit passing "h" for the help in addition to "help".

jonatack committed:
netinfo: clarify relaytxes and addr_relay_enabled help docs

jonatack committed:
netinfo: add peer services column

tdb3 reviewed (commented):
Approach ACK

Nice improvement.
Performed a quick initial code review and exercised the changes interactively running on signet. I plan to circle back.

tdb3 commented on src/bitcoin-cli.cpp:
```diff
@@ -456,6 +458,16 @@ class NetinfoRequestHandler : public BaseRequestHandler
+    {
+        std::string str;
+        for (size_t i = 0; i < services.size(); ++i) {
+            if (services[i].isNull()) continue;
```
Do we expect to receive a null in `servicesnames` among other entries?
If we don't, maybe this should be something like `assert(!services[i].isNull());` so the issue can be noticed.

tdb3 replied:
Sorry, I wasn't as clear as I could have been.  It seems possible to have no services (e.g. `servicesnames` has no elements, in which case the for loop wouldn't be entered, right?).  What I meant to ask is if we think we'll see `servicesnames` with some non-zero number of elements, with at least one element satisfying the check to `isNull()`?  In other words, does the `continue` ever actually happen?

jonatack replied:
Looking at it more, I don't think it's needed and won't be hit, so removed that line for now. (Thanks!)

tdb3 reviewed (commented):
Left a couple small comments.
Tested that using `5` does include only outbound peers by connecting two nodes locally.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -391,9 +391,10 @@ class NetinfoRequestHandler : public BaseRequestHandler
+    bool OutboundOnlySelected() const { return m_details_level == 5; }
+    bool DetailsRequested() const { return m_details_level > 0 && m_details_level <= MAX_DETAIL_LEVEL; }
+    bool IsAddressSelected() const { return m_details_level == 2 || m_details_level >= 4; }
+    bool IsVersionSelected() const { return m_details_level == 3 || m_details_level >= 4; }
```
done

jonatack commented:
Updated to take @tdb3 review feedback (thanks!) and few adjustments.

<details><summary><code>git diff 51206c6 683b558</code></summary><p>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index 3207a5d9c97..000f7695de6 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -391,10 +391,10 @@ private:
         return UNKNOWN_NETWORK;
     }
     uint8_t m_details_level{0}; //!< Optional user-supplied arg to set dashboard details level
-    bool OutboundOnlySelected() const { return m_details_level == 5; }
     bool DetailsRequested() const { return m_details_level > 0 && m_details_level <= MAX_DETAIL_LEVEL; }
     bool IsAddressSelected() const { return m_details_level == 2 || m_details_level >= 4; }
-    bool IsVersionSelected() const { return m_details_level == 3 || m_details_level >= 4; }
+    bool IsVersionSelected() const { return m_details_level >= 3; }
+    bool OutboundOnlySelected() const { return m_details_level == 5; }
     bool m_is_asmap_on{false};
     size_t m_max_addr_length{0};
     size_t m_max_addr_processed_length{5};
@@ -407,8 +407,8 @@ private:
         std::string sub_version;
         std::string conn_type;
         std::string network;
-        std::string services;
         std::string age;
+        std::string services;
         std::string transport_protocol_type;
         double min_ping;
         double ping;
@@ -463,7 +463,6 @@ private:
     {
         std::string str;
         for (size_t i = 0; i < services.size(); ++i) {
-            if (services[i].isNull()) continue;
             const std::string s{services[i].get_str()};
             str += s == "NETWORK_LIMITED" ? 'l' : s == "P2P_V2" ? '2' : ToLower(s[0]);
         }
@@ -539,7 +538,7 @@ public:
                 const bool is_addr_relay_enabled{peer["addr_relay_enabled"].isNull() ? false : peer["addr_relay_enabled"].get_bool()};
                 const bool is_bip152_hb_from{peer["bip152_hb_from"].get_bool()};
                 const bool is_bip152_hb_to{peer["bip152_hb_to"].get_bool()};
-                m_peers.push_back({addr, sub_version, conn_type, NETWORK_SHORT_NAMES[network_id], services, age, transport, min_ping, ping, addr_processed, addr_rate_limited, last_blck, last_recv, last_send, last_trxn, peer_id, mapped_as, version, is_addr_relay_enabled, is_bip152_hb_from, is_bip152_hb_to, is_outbound, is_tx_relay});
+                m_peers.push_back({addr, sub_version, conn_type, NETWORK_SHORT_NAMES[network_id], age, services, transport, min_ping, ping, addr_processed, addr_rate_limited, last_blck, last_recv, last_send, last_trxn, peer_id, mapped_as, version, is_addr_relay_enabled, is_bip152_hb_from, is_bip152_hb_to, is_outbound, is_tx_relay});
                 m_max_addr_length = std::max(addr.length() + 1, m_max_addr_length);
                 m_max_addr_processed_length = std::max(ToString(addr_processed).length(), m_max_addr_processed_length);
                 m_max_addr_rate_limited_length = std::max(ToString(addr_rate_limited).length(), m_max_addr_rate_limited_length);
@@ -684,13 +683,13 @@ public:
         "           \"addr\"   - address fetch; short-lived connection for requesting addresses\n"
         "  net      Network the peer connected through (\"ipv4\", \"ipv6\", \"onion\", \"i2p\", \"cjdns\", or \"npr\" (not publicly routable))\n"
         "  serv     Services offered by the peer\n"
-        "           \"n\" - NETWORK, can serve the full block chain\n"
-        "           \"b\" - BLOOM, can handle bloom-filtered connections (see BIP 111)\n"
-        "           \"w\" - WITNESS, can be asked for blocks and transactions with witness data (SegWit)\n"
-        "           \"c\" - COMPACT_FILTERS, can handle basic block filter requests (see BIPs 157 and 158)\n"
-        "           \"l\" - NETWORK_LIMITED, limited to serving only the last 288 blocks (~2 days)\n"
-        "           \"2\" - P2P_V2, supports version 2 P2P transport protocol, as defined in BIP324\n"
-        "           \"u\" - UNKNOWN, unrecognized bit flag\n"
+        "           \"n\" - NETWORK: peer can serve the full block chain\n"
+        "           \"b\" - BLOOM: peer can handle bloom-filtered connections (see BIP 111)\n"
+        "           \"w\" - WITNESS: peer can be asked for blocks and transactions with witness data (SegWit)\n"
+        "           \"c\" - COMPACT_FILTERS: peer can handle basic block filter requests (see BIPs 157 and 158)\n"
+        "           \"l\" - NETWORK_LIMITED: peer limited to serving only the last 288 blocks (~2 days)\n"
+        "           \"2\" - P2P_V2: peer supports version 2 P2P transport protocol, as defined in BIP 324\n"
+        "           \"u\" - UNKNOWN: unrecognized bit flag\n"
         "  v        Version of transport protocol used for the connection\n"
         "  mping    Minimum observed ping time, in milliseconds (ms)\n"
         "  ping     Last observed ping time, in milliseconds (ms)\n"
```

</p></details>

tdb3 reviewed (approved):
code review and light test ACK 683b558a020f1632b0a3cbdaa165adbd1423281c

Good catch with age/services

1440000bytes reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/30930/commits/683b558a020f1632b0a3cbdaa165adbd1423281c

vasild commented on src/bitcoin-cli.cpp:
```diff
@@ -648,6 +666,7 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "                                  2 - Like 1 but with an address column\n"
         "                                  3 - Like 1 but with a version column\n"
         "                                  4 - Like 1 but with both address and version columns\n"
+        "                                  5 - Like 4 but limited to outbound peers only, to save screen space\n"
```
Before these numbers somehow designated increasing verbosity and more columns. Whereas now the newly added "5" does not add more columns, but _removes rows_ compared to "4".

I am not sure how to improve that. It is a different dimension, maybe use a different option "outonly". Then one could have either one of the 0-4 modes in "outonly" mode or in full mode (in + out).

(non-blocker, just something to consider)

jonatack replied:
I see, but think it's simpler and more convenient for the user (i.e. me) to type 5 rather than 4, or vice versa.

I did maintain the "high details levels = maximum" behavior preferred by @laanwj in https://github.com/bitcoin/bitcoin/pull/21192 for values above 5.

jonatack replied:
Also, I find myself only using level 4 or 5. Perhaps other reviewers would prefer the option to apply, as you suggest, to all the other levels. Happy to do that (here or in a follow-up) if people prefer.

jonatack replied:
Updated to take @vasild's proposal. I believe it's the best path of the 3 options discussed above.

laanwj replied:
> I did maintain the "high details levels = maximum" behavior preferred by laanwj in https://github.com/bitcoin/bitcoin/pull/21192 for values above 5.

Thank you 😄

vasild reviewed (approved):
ACK 683b558a020f1632b0a3cbdaa165adbd1423281c

LarryRuane reviewed (changes requested):
I think you also should update line 93 (change "4" to "5")
```
argsman.AddArg("-netinfo", "Get network peer connection information from the remote server. An optional integer argument from 0 to 4 can be passed for different peers listings (default: 0). Pass \"help\" for detailed help documentation.", ArgsManager::ALLOW_ANY, OptionsCategory::CLI_COMMANDS);
```

jonatack commented:
> I think you also should update line 93 (change "4" to "5")

Well spotted 👍 I'm thinking of taking @vasild's outonly suggestion, and will fix that if not.

jonatack committed:
netinfo: rename and hoist max level constant to use in top-level help

to avoid overlooking to update the top-level help if the value of the constant changes.

jonatack committed:
netinfo: allow setting an outbound-only peer list

by passing an additional argument of "outonly" or "o".

This has been requested in order to keep the output within screen limits when running -netinfo
as a live dashboard, i.e. with `watch`.

Also allow passing "h" in addition to "help" to see the help documentation.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -456,6 +459,15 @@ class NetinfoRequestHandler : public BaseRequestHandler
+        std::string str;
+        for (size_t i = 0; i < services.size(); ++i) {
+            const std::string s{services[i].get_str()};
+            str += s == "NETWORK_LIMITED" ? 'l' : s == "P2P_V2" ? '2' : ToLower(s[0]);
```
Didn't take, as those braces are unneeded due to the low precedence of the ternary operator, clang-format doesn't prefer them, and they add clutter. Will add if people insist.

jonatack commented:
> I think you also should update line 93 (change "4" to "5")
> 
> ```
> argsman.AddArg("-netinfo", "Get network peer connection information from the remote server. An optional integer argument from 0 to 4 can be passed for different peers listings (default: 0). Pass \"help\" for detailed help documentation.", ArgsManager::ALLOW_ANY, OptionsCategory::CLI_COMMANDS);
> ```

Good catch @LarryRuane -- added 153b898a77f4c892731c220ce85d29c1cf880355 to ensure we don't overlook updating this.

jonatack commented:
Updated per `git diff 683b558 1847b9` to take @vasild's [suggestion](https://github.com/bitcoin/bitcoin/pull/30930#discussion_r1795748339) and a fix for @LarryRuane's [catch](https://github.com/bitcoin/bitcoin/pull/30930#pullrequestreview-2363290423) (thanks!)

jonatack commented:
Updated to take @vasild's [suggestion](https://github.com/bitcoin/bitcoin/pull/30930#discussion_r1795748339) and a fix for @LarryRuane's [catch](https://github.com/bitcoin/bitcoin/pull/30930#pullrequestreview-2363290423) (thanks!) and doc updates I overlooked.

<details><summary><code>git diff 683b558 17f8f03</code></summary><p>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index 000f7695de6..6bda4cbe0c5 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -57,6 +57,7 @@ static const int DEFAULT_HTTP_CLIENT_TIMEOUT=900;
 static constexpr int DEFAULT_WAIT_CLIENT_TIMEOUT = 0;
 static const bool DEFAULT_NAMED=false;
 static const int CONTINUE_EXECUTION=-1;
+static constexpr uint8_t NETINFO_MAX_LEVEL{4};
 static constexpr int8_t UNKNOWN_NETWORK{-1};
 // See GetNetworkName() in netbase.cpp
 static constexpr std::array NETWORKS{"not_publicly_routable", "ipv4", "ipv6", "onion", "i2p", "cjdns", "internal"};
@@ -90,7 +91,7 @@ static void SetupCliArgs(ArgsManager& argsman)
                    ArgsManager::ALLOW_ANY, OptionsCategory::CLI_COMMANDS);
     argsman.AddArg("-addrinfo", "Get the number of addresses known to the node, per network and total, after filtering for quality and recency. The total number of addresses known to the node may be higher.", ArgsManager::ALLOW_ANY, OptionsCategory::CLI_COMMANDS);
     argsman.AddArg("-getinfo", "Get general information from the remote server. Note that unlike server-side RPC calls, the output of -getinfo is the result of multiple non-atomic requests. Some entries in the output may represent results from different states (e.g. wallet balance may be as of a different block from the chain state reported)", ArgsManager::ALLOW_ANY, OptionsCategory::CLI_COMMANDS);
-    argsman.AddArg("-netinfo", "Get network peer connection information from the remote server. An optional integer argument from 0 to 4 can be passed for different peers listings (default: 0). Pass \"help\" for detailed help documentation.", ArgsManager::ALLOW_ANY, OptionsCategory::CLI_COMMANDS);
+    argsman.AddArg("-netinfo", strprintf("Get network peer connection information from the remote server. An optional argument from 0 to %d can be passed for different peers listings (default: 0). If a non-zero value is passed, an additional \"outonly\" (or \"o\") argument can be passed to see outbound peers only. Pass \"help\" (or \"h\") for detailed help documentation.", NETINFO_MAX_LEVEL), ArgsManager::ALLOW_ANY, OptionsCategory::CLI_COMMANDS);
 
     SetupChainParamsBaseOptions(argsman);
     argsman.AddArg("-color=<when>", strprintf("Color setting for CLI output (default: %s). Valid values: always, auto (add color codes when standard output is connected to a terminal and OS is not WIN32), never.", DEFAULT_COLOR_SETTING), ArgsManager::ALLOW_ANY | ArgsManager::DISALLOW_NEGATION, OptionsCategory::OPTIONS);
@@ -379,7 +380,6 @@ public:
 class NetinfoRequestHandler : public BaseRequestHandler
 {
 private:
-    static constexpr uint8_t MAX_DETAIL_LEVEL{5};
     std::array<std::array<uint16_t, NETWORKS.size() + 1>, 3> m_counts{{{}}}; //!< Peer counts by (in/out/total, networks/total)
     uint8_t m_block_relay_peers_count{0};
     uint8_t m_manual_peers_count{0};
@@ -391,10 +391,10 @@ private:
         return UNKNOWN_NETWORK;
     }
     uint8_t m_details_level{0}; //!< Optional user-supplied arg to set dashboard details level
-    bool DetailsRequested() const { return m_details_level > 0 && m_details_level <= MAX_DETAIL_LEVEL; }
-    bool IsAddressSelected() const { return m_details_level == 2 || m_details_level >= 4; }
-    bool IsVersionSelected() const { return m_details_level >= 3; }
-    bool OutboundOnlySelected() const { return m_details_level == 5; }
+    bool DetailsRequested() const { return m_details_level > 0 && m_details_level < 5; }
+    bool IsAddressSelected() const { return m_details_level == 2 || m_details_level == 4; }
+    bool IsVersionSelected() const { return m_details_level == 3 || m_details_level == 4; }
+    bool m_outbound_only_selected{false};
     bool m_is_asmap_on{false};
     size_t m_max_addr_length{0};
     size_t m_max_addr_processed_length{5};
@@ -478,9 +478,18 @@ public:
         if (!args.empty()) {
             uint8_t n{0};
             if (ParseUInt8(args.at(0), &n)) {
-                m_details_level = (n <= MAX_DETAIL_LEVEL) ? n : 4;
+                m_details_level = std::min(n, NETINFO_MAX_LEVEL);
             } else {
-                throw std::runtime_error(strprintf("invalid -netinfo argument: %s\nFor more information, run: bitcoin-cli -netinfo help", args.at(0)));
+                throw std::runtime_error(strprintf("invalid -netinfo level argument: %s\nFor more information, run: bitcoin-cli -netinfo help", args.at(0)));
+            }
+            if (args.size() > 1) {
+                if (std::string_view s{args.at(1)}; n && (s == "o" || s == "outonly")) {
+                    m_outbound_only_selected = true;
+                } else if (n) {
+                    throw std::runtime_error(strprintf("invalid -netinfo outonly argument: %s\nFor more information, run: bitcoin-cli -netinfo help", args.at(1)));
+                } else {
+                    throw std::runtime_error(strprintf("invalid -netinfo outonly argument: %s\nThe outonly argument is only valid for details levels higher than 0. For more information, run: bitcoin-cli -netinfo help", args.at(1)));
+                }
             }
         }
         UniValue result(UniValue::VARR);
@@ -515,7 +524,7 @@ public:
             ++m_counts.at(2).at(NETWORKS.size());           // total overall
             if (conn_type == "block-relay-only") ++m_block_relay_peers_count;
             if (conn_type == "manual") ++m_manual_peers_count;
-            if (!is_outbound && OutboundOnlySelected()) continue;
+            if (m_outbound_only_selected && !is_outbound) continue;
             if (DetailsRequested()) {
                 // Push data for this peer to the peers vector.
                 const int peer_id{peer["id"].getInt<int>()};
@@ -650,26 +659,28 @@ public:
     }
 
     const std::string m_help_doc{
-        "-netinfo level|\"help\" \n\n"
+        "-netinfo level|outonly|help\n\n"
         "Returns a network peer connections dashboard with information from the remote server.\n"
         "This human-readable interface will change regularly and is not intended to be a stable API.\n"
         "Under the hood, -netinfo fetches the data by calling getpeerinfo and getnetworkinfo.\n"
-        + strprintf("An optional integer argument from 0 to %d can be passed for different peers listings; %d to 255 are parsed as %d.\n", MAX_DETAIL_LEVEL, MAX_DETAIL_LEVEL, MAX_DETAIL_LEVEL) +
-        "Pass \"help\" to see this detailed help documentation.\n"
-        "If more than one argument is passed, only the first one is read and parsed.\n"
-        "Suggestion: use with the Linux watch(1) command for a live dashboard; see example below.\n\n"
+        + strprintf("An optional argument from 0 to %d can be passed for different peers listings; values above %d up to 255 are parsed as %d.\n", NETINFO_MAX_LEVEL, NETINFO_MAX_LEVEL, NETINFO_MAX_LEVEL) +
+        "If that argument is passed, an optional additional \"outonly\" argument may be passed to obtain the listing with outbound peers only.\n"
+        "Pass \"help\" or \"h\" to see this detailed help documentation.\n"
+        "If more than two arguments are passed, only the first two are read and parsed.\n"
+        "Suggestion: use -netinfo with the Linux watch(1) command for a live dashboard; see example below.\n\n"
         "Arguments:\n"
-        + strprintf("1. level (integer 0-%d, optional)  Specify the info level of the peers dashboard (default 0):\n", MAX_DETAIL_LEVEL) +
+        + strprintf("1. level (integer 0-%d, optional)  Specify the info level of the peers dashboard (default 0):\n", NETINFO_MAX_LEVEL) +
         "                                  0 - Peer counts for each reachable network as well as for block relay peers\n"
         "                                      and manual peers, and the list of local addresses and ports\n"
         "                                  1 - Like 0 but preceded by a peers listing (without address and version columns)\n"
         "                                  2 - Like 1 but with an address column\n"
         "                                  3 - Like 1 but with a version column\n"
         "                                  4 - Like 1 but with both address and version columns\n"
-        "                                  5 - Like 4 but limited to outbound peers only, to save screen space\n"
-        "2. help (string \"help\", optional) Print this help documentation instead of the dashboard.\n\n"
+        "2. outonly (\"outonly\" or \"o\", optional) Return the peers listing with outbound peers only, i.e. to save screen space\n"
+        "                                        when a node has many inbound peers. Only valid if a level is passed.\n\n"
+        "help (\"help\" or \"h\", optional) Print this help documentation instead of the dashboard.\n\n"
         "Result:\n\n"
-        + strprintf("* The peers listing in levels 1-%d displays all of the peers sorted by direction and minimum ping time:\n\n", MAX_DETAIL_LEVEL) +
+        + strprintf("* The peers listing in levels 1-%d displays all of the peers sorted by direction and minimum ping time:\n\n", NETINFO_MAX_LEVEL) +
         "  Column   Description\n"
         "  ------   -----------\n"
         "  <->      Direction\n"
@@ -719,9 +730,11 @@ public:
         "The same, preceded by a peers listing without address and version columns\n"
         "> bitcoin-cli -netinfo 1\n\n"
         "Full dashboard\n"
-        + strprintf("> bitcoin-cli -netinfo %d\n\n", MAX_DETAIL_LEVEL) +
+        + strprintf("> bitcoin-cli -netinfo %d\n\n", NETINFO_MAX_LEVEL) +
+        "Full dashboard, but with outbound peers only\n"
+        + strprintf("> bitcoin-cli -netinfo %d outonly\n\n", NETINFO_MAX_LEVEL) +
         "Full live dashboard, adjust --interval or --no-title as needed (Linux)\n"
-        + strprintf("> watch --interval 1 --no-title bitcoin-cli -netinfo %d\n\n", MAX_DETAIL_LEVEL) +
+        + strprintf("> watch --interval 1 --no-title bitcoin-cli -netinfo %d\n\n", NETINFO_MAX_LEVEL) +
         "See this help\n"
         "> bitcoin-cli -netinfo help\n"};
 };
@@ -1245,7 +1258,7 @@ static int CommandLineRPC(int argc, char *argv[])
         if (gArgs.IsArgSet("-getinfo")) {
             rh.reset(new GetinfoRequestHandler());
         } else if (gArgs.GetBoolArg("-netinfo", false)) {
-            if (!args.empty() && args.at(0) == "help") {
+            if (!args.empty() && (args.at(0) == "h" || args.at(0) == "help")) {
                 tfm::format(std::cout, "%s\n", NetinfoRequestHandler().m_help_doc);
                 return 0;
             }
```
</p></details>

1440000bytes reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/30930/commits/17f8f03ec69b6dd3d61137a39b8f88201ec500dc

Changes since last review: https://github.com/bitcoin/bitcoin/pull/30930#issuecomment-2420629857

tdb3 reviewed (approved):
re ACK 17f8f03ec69b6dd3d61137a39b8f88201ec500dc

Code reviewed diff.

Ran sanity check using the new `outonly` arg:
```
$ build/src/bitcoin-cli -signet -netinfo 1
Bitcoin Core client v28.99.0-17f8f03ec69b signet - server 70016/Satoshi:28.99.0/

<->   type   net   serv  v  mping   ping send recv  txn  blk  hb addrp addrl  age id 
 in          npr   nwl2  2      1      2   38   24    1              2          6 13 
out   full  ipv6    nwl  1     82     91   24   24         3  .   1001         25  6 
...

$ build/src/bitcoin-cli -signet -netinfo 1 outonly
Bitcoin Core client v28.99.0-17f8f03ec69b signet - server 70016/Satoshi:28.99.0/

<->   type   net   serv  v  mping   ping send recv  txn  blk  hb addrp addrl  age id 
out   full  ipv6    nwl  1     82     91   26   26         3  .   1001         25  6 
...
```

Compared the output of 17f8f03ec69b6dd3d61137a39b8f88201ec500dc to master (e8f72aefd20049eac81b150e7f0d33709acd18ed) for 1 through 4.  Saw `serv` column and no unexpected differences. 

Sanity checked that >4 acts as 4.

vasild commented on src/bitcoin-cli.cpp:
```diff
@@ -466,9 +478,18 @@ class NetinfoRequestHandler : public BaseRequestHandler
+                throw std::runtime_error(strprintf("invalid -netinfo level argument: %s\nFor more information, run: bitcoin-cli -netinfo help", args.at(0)));
+            }
+            if (args.size() > 1) {
+                if (std::string_view s{args.at(1)}; n && (s == "o" || s == "outonly")) {
+                    m_outbound_only_selected = true;
+                } else if (n) {
+                    throw std::runtime_error(strprintf("invalid -netinfo outonly argument: %s\nFor more information, run: bitcoin-cli -netinfo help", args.at(1)));
+                } else {
+                    throw std::runtime_error(strprintf("invalid -netinfo outonly argument: %s\nThe outonly argument is only valid for details levels higher than 0. For more information, run: bitcoin-cli -netinfo help", args.at(1)));
```
nit: since `args.at(1)` is repeated also in the `throw` statements, maybe:

```cpp
std::string_view arg1{args.at(1)};
if (n && (arg1 == "o" || arg1 == "outonly")) {
} else {
    throw ... arg1 ...
```

jonatack replied:
done

vasild commented on src/bitcoin-cli.cpp:
```diff
@@ -632,25 +659,28 @@ class NetinfoRequestHandler : public BaseRequestHandler
 
     const std::string m_help_doc{
-        "-netinfo level|\"help\" \n\n"
+        "-netinfo level|outonly|help\n\n"
```
nit: `level|outonly|help` reads as "exactly one of `level`, `outonly` or `help`". I guess it should be

```
-netinfo (level [outonly]) | help
```
or two lines:

```
        "-netinfo \"help\"\n"
        "-netinfo level [outonly]\n\n"
```

jonatack replied:
Thanks! Done.

vasild reviewed (approved):
ACK 17f8f03ec69b6dd3d61137a39b8f88201ec500dc

jonatack commented:
Updated the last commit to take @vasild's feedback (thanks!) per the following small diff.

<details><summary><code>git diff 17f8f03 87532fe</code></summary><p>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index 6bda4cbe0c5..f63dd22e621 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -486,9 +486,9 @@ public:
                 if (std::string_view s{args.at(1)}; n && (s == "o" || s == "outonly")) {
                     m_outbound_only_selected = true;
                 } else if (n) {
-                    throw std::runtime_error(strprintf("invalid -netinfo outonly argument: %s\nFor more information, run: bitcoin-cli -netinfo help", args.at(1)));
+                    throw std::runtime_error(strprintf("invalid -netinfo outonly argument: %s\nFor more information, run: bitcoin-cli -netinfo help", s));
                 } else {
-                    throw std::runtime_error(strprintf("invalid -netinfo outonly argument: %s\nThe outonly argument is only valid for details levels higher than 0. For more information, run: bitcoin-cli -netinfo help", args.at(1)));
+                    throw std::runtime_error(strprintf("invalid -netinfo outonly argument: %s\nThe outonly argument is only valid for a level greater than 0 (the first argument). For more information, run: bitcoin-cli -netinfo help", s));
                 }
             }
         }
@@ -659,7 +659,7 @@ public:
     }
 
     const std::string m_help_doc{
-        "-netinfo level|outonly|help\n\n"
+        "-netinfo (level [outonly]) | help\n\n"
         "Returns a network peer connections dashboard with information from the remote server.\n"
```
</p></details>

rkrux commented on src/bitcoin-cli.cpp:
```diff
@@ -456,6 +458,15 @@ class NetinfoRequestHandler : public BaseRequestHandler
+    {
+        std::string str;
+        for (size_t i = 0; i < services.size(); ++i) {
+            const std::string s{services[i].get_str()};
```
Can `NODE_NONE` ever be a value here?

https://github.com/bitcoin/bitcoin/blob/master/src/protocol.h#L312

jonatack replied:
I think so. Using this, I see empty service names while a peer connection is setting up, and often from one or two (usually inbound tor) peers. It looks like `NODE_NONE` would be returned as an empty string by `serviceFlagsToStr()`.

rkrux replied:
Thanks for clarifying, I was wondering whether the below line would cause `NODE_NONE` to show up as `n` but that should not be the case.

rkrux reviewed (approved):
tACK 87532fe55856efc063cf81244800da37a015ba75

Successful make and functional tests. I used the `-netinfo` command with various verbosities and the `outonly` option. Also checked for verbosities greater than 4.

Asked a question for clarity.

tdb3 reviewed (approved):
cr re ACK 87532fe55856efc063cf81244800da37a015ba75

brunoerg reviewed (commented):
crACK 87532fe55856efc063cf81244800da37a015ba75

code looks great, but I didn't test in practice yet.

achow101 commented:
ACK 87532fe55856efc063cf81244800da37a015ba75
